### PR TITLE
JVM_IR: remove most uses of StackValue from ExpressionCodegen

### PR DESCRIPTION
**Pull request by pyos**
Part 2 (cont. of #2229)

### CODE REVIEW

**Comment by pyos** [comment]
@max-kammerer I've made a couple changes to make this easier to merge with the inlining: removed a bytecode-changing commit, left `irFunction` and `frame` public (although the latter is also available as `frameMap`...) and did not remove `genFinallyBlock` so you have a place to put the marker method calls. Merging should now boil down to combining changes to visitReturn and using unwindBlockStack instead of generateFinallyBlocksIfNeeded. I've checked the tests and it seems like there are no new failing ones now (commit 7 of the previous review only affected one bytecode text test because it did not use genFinallyBlock for one of the copies; the inliner doesn't actually need that one, but I've left it in anyway).

**Comment by max-kammerer** [comment]
@pyos please rebase on master there is conflict with #2257

**Comment by pyos** [comment]
@max-kammerer done

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt [hunk 1]
```diff
@@ -104,26 +100,33 @@ class BlockInfo private constructor(val parent: BlockInfo?) {
 
 class VariableInfo(val declaration: IrVariable, val index: Int, val type: Type, val startLabel: Label)
 
-@Suppress("IMPLICIT_CAST_TO_ANY")
 class ExpressionCodegen(
     val irFunction: IrFunction,
     val frame: IrFrameMap,
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt]: val frame ->  override val frameMap,
and next frameMap declaration would be redundant

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt [hunk 2]
```diff
@@ -183,36 +176,23 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
...
                 return null
 
-            genSubject()
             genSwitch()
-            genThenExpressions()
-            val result = genElseExpression()
-
-            mv.mark(endLabel)
-            return result
+            return genBranchTargets()
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: @ting-yuan Please review changes in SwitchGenerator.kt, as I understand there should be several more specific generators and method merging could break your plans/design
```
            genSubject()
            genSwitch()
            genThenExpressions()
```

**Comment by max-kammerer** [comment]
@pyos I would prefer to keep 'gaps' logic in one place, so please rollback "Partially inline genFinallyBlock pyos"

**Comment by pyos** [comment]
@max-kammerer done. I guess there's not much point in that if a redundant copy of `finally` cannot be removed because of the inliner, yeah.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt (hunk 2 above)
Comment (ting-yuan) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt]: Although I prefer smaller functions as they are easier to read, this looks fine to me.

**Comment by max-kammerer** [comment]
@pyos Please rebase last time and I will merge pr: github warns about conflict

**Comment by max-kammerer** [comment]
@pyos Sorry, I missed that you updated last commit. I would prefer to distinguish NewArray and ArrayConstructors: maybe with more proper naming. At least last commit could be postponed to another PR

**Comment by pyos** [comment]
@max-kammerer ok, I'll submit that separately.

>Please rebase last time and I will merge pr: github warns about conflict

Weird, it said there were no conflicts, and git pull --rebase completed automatically.

**Comment by max-kammerer** [comment]
@pyos Thank you! Nice work!

**Pull request merged**